src/msg/producer/writer/message.go

```diff
@@ -38,22 +41,22 @@ type message struct {
 	retried      int
 	// NB(cw) isAcked could be accessed concurrently by the background thread
 	// in message writer and acked by consumer service writers.
-	isAcked *atomic.Bool
+	// Safe to store value inside struct, as message is never copied by value
+	isAcked atomic.Bool
 }
 
 func newMessage() *message {
 	return &message{
 		retryAtNanos: 0,
 		retried:      0,
-		isAcked:      atomic.NewBool(false),
 	}
 }
 
 // Set sets the message.
 func (m *message) Set(meta metadata, rm *producer.RefCountedMessage, initNanos int64) {
 	m.initNanos = initNanos
 	m.meta = meta
-	m.RefCountedMessage = rm
+	m.storeRefCountedMessage(rm)
```

Comment:
When would `Set()` and `Ack()` be called concurrently? I remember `Set` is mainly used for pooling, in the lifecycle of a message object, `Set` and `Ack` could never be called concurrently?

Reply:
so, race detector is actually complaining about Set() being called in another goroutine from Ack(), it's not aware of message reuse due to pooling.